cmd/snap-bootstrap/partition/sfdisk.go

```diff
@@ -270,36 +314,27 @@ func deviceLayoutFromPartitionTable(ptable sfdiskPartitionTable) (*DeviceLayout,
 		}
 		bd := info.BlockDevices[0]
 
+		vsType, err := fromSfdiskPartitionType(p.Type, ptable.Label)
+		if err != nil {
+			return nil, fmt.Errorf("cannot convert sfdisk partition type %q: %v", p.Type, err)
+		}
+
 		structure[i] = gadget.VolumeStructure{
 			Name:       p.Name,
 			Size:       gadget.Size(p.Size) * sectorSize,
 			Label:      bd.Label,
-			Type:       p.Type,
+			Type:       vsType,
 			Filesystem: bd.FSType,
 		}
 
-		// sometimes sfdisk reports what is "0C" in gadget.yaml as "c", so fix
-		// that up now
-		// TODO:UC20: what's a better way?
-		if structure[i].Type == "c" {
-			structure[i].Type = "0C"
-		}
-
-		// sometimes sfdisk reports no "Name" for partitions, so in this case
-		// if we also have a label for the partition from lsblk, we should
-		// fallback to that instead
-		if structure[i].Name == "" && structure[i].Label != "" {
-			structure[i].Name = structure[i].Label
-		}
```

Comment:
@bboozzoo this part is still necessary, as-is with the branch we don't have compatible layouts for the disk + gadget because the ensureLayoutCompatibility check only compares dv.Name == gv.Name, and without this snippet, the dv will not have a Name.

Reply:
I see. Hm that should be unnecessary though. The check in bootstrap.go needs to be fixed as it's based on wrong assuptions clearly. You can't have a name in MBR. At the same time, the gadget is free to specify a different and name and fs label regardless of schema, so promoting the label to be the name does not ensure success.

Reply:
well I adapted the check slightly like this:

```patch
Author: Ian Johnson <ian.johnson@canonical.com>
Date:   Fri Apr 17 14:06:03 2020 -0500

    cmd/snap-bootstrap/bootstrap: also use name == label when ensuring compatibility
    
    This is needed since we no longer will set the Name of the device structure from
    the device structure's label, so we also need to check the possibility that the
    device structure's label matches the gadget structure's name.
    
    Signed-off-by: Ian Johnson <ian.johnson@canonical.com>

diff --git a/cmd/snap-bootstrap/bootstrap/bootstrap.go b/cmd/snap-bootstrap/bootstrap/bootstrap.go
index 4bec579865..2d43c21f6e 100644
--- a/cmd/snap-bootstrap/bootstrap/bootstrap.go
+++ b/cmd/snap-bootstrap/bootstrap/bootstrap.go
@@ -257,8 +257,14 @@ func ensureLayoutCompatibility(gadgetLayout *gadget.LaidOutVolume, diskLayout *p
        eq := func(ds partition.DeviceStructure, gs gadget.LaidOutStructure) bool {
                dv := ds.VolumeStructure
                gv := gs.VolumeStructure
-               // Previous installation may have failed before filesystem creation or partition may be encrypted
-               check := dv.Name == gv.Name && ds.StartOffset == gs.StartOffset && (ds.Created || dv.Filesystem == gv.Filesystem)
+
+               // the name may not exist on the device structure if this is an MBR
+               // disk, in which case we have to compare the gadget structure name
+               // with the disk structure label
+               check := (dv.Name == gv.Name || dv.Label == gv.Name) &&
+                       ds.StartOffset == gs.StartOffset &&
+                       // Previous installation may have failed before filesystem creation or partition may be encrypted
+                       (ds.Created || dv.Filesystem == gv.Filesystem)
 
                if gv.Role == gadget.SystemData {
                        // system-data may have been expanded
```

Reply:
As discussed on IRC, https://github.com/snapcore/snapd/pull/8487/commits/e8629278382af1170ea3f1d4351b00ed7107121d contains the fix.